jumanji/environments/routing/rware/env.py

```diff
@@ -0,0 +1,527 @@
+import jax.numpy as jnp
+import matplotlib
+
+from jumanji import specs
+from jumanji.env import Environment
+from jumanji.environments.routing.rware import utils
+from jumanji.environments.routing.rware.generator import Generator, RandomGenerator
+from jumanji.environments.routing.rware.types import (
+    _SHELVES,
+    Action,
+    Agent,
+    Direction,
+    Observation,
+    Shelf,
+    State,
+)
+from jumanji.environments.routing.rware.viewer import RwareViewer
+from jumanji.tree_utils import tree_slice
+from jumanji.types import TimeStep, restart, transition, truncation
+from jumanji.viewer import Viewer
+
+
+class Rware(Environment[State]):
+    """A JAX implementation of the 'Robotic warehouse' environment:
+    https://github.com/semitable/robotic-warehouse
+    which is described in the paper [1].
+
+    Creates a grid world where multiple agents (robots)
+    are supposed to collect shelves, bring them to a goal
+    and then return them.
+
+    Below is an example warehouse floor grid:
+    the grid layout is instantiated using three arguments
+
+    - shelf_rows: number of vertical shelf clusters
+    - shelf_columns: odd number of horizontal shelf clusters
+    - column_height: height of each cluster
+
+    A cluster is a set of grouped shelves (two cells wide) represented
+    below as
+
+                        XX
+    Shelf cluster ->    XX    (this cluster is of height 3)
+                        XX
+
+    Grid Layout:
+
+                   shelf columns (here set to 3, i.e.
+                     v  v  v      shelf_columns=3, must be an odd number)
+                    ----------
+                 >  -XX-XX-XX-        ^
+    Shelf Row 1 ->  -XX-XX-XX-  Column Height (here set to 3, i.e.
+                 >  -XX-XX-XX-        v        column_height=3)
+                    ----------
+                    -XX----XX-   <
+                    -XX----XX-   <- Shelf Row 2 (here set to 2, i.e.
+                    -XX----XX-   <              shelf_rows=2)
+                    ----------
+                    ----GG----
+
+    - G: is the goal positions where agents are rewarded if
+    they successfully deliver a requested shelf (i.e toggle the load action
+    inside the goal position while carrying a requested shelf).
+
+    The final grid size will be
+    - height: (column_height + 1) * shelf_rows + 2
+    - width: (2 + 1) * shelf_columns + 1
+
+    The bottom-middle column is removed to allow for
+    agents to queue in front of the goal positions
+
+    - action: jax array (int) of shape (num_agents,) containing the action for each agent.
+        (0: noop, 1: forward, 2: left, 3: right, 4: toggle_load)
+
+    - reward: jax array (int) of shape (), global reward shared by all agents, +1
+        for every successful delivery of a requested shelf to the goal position.
+
+    - episode termination:
+        - The number of steps is greater than the limit.
+        - Any agent selects an action which causes two agents to collide.
```

Comment:
I'm sure this has been discussed, but why not just make one of them no-op?

Reply:
Early on @clement-bonnet said this was the preferred behaviour.